### PR TITLE
RP2: Add support for Arduino Nano RP2040

### LINKED ISSUE #7634: [Board/Arduino Nano RP2040 Connect] Adding the first rp2 board with wifi.
Hey could someone advice me of getting this board onboarded with micropython?

I wrote me an neopixel implementation but now I am struggeling with the missing WIFI.

That's what I found in the circuitpython project. But sadly circuitpython does not support `uasyncio`.
```c
USB_VID = 0x239A
USB_PID = 0x00CF

USB_PRODUCT = "Arduino Nano RP2040 Connect"
USB_MANUFACTURER = "Arduino"

CHIP_VARIANT = RP2040
CHIP_FAMILY = rp2

EXTERNAL_FLASH_DEVICES = "AT25SF128A, IS25LP128F"

CIRCUITPY__EVE = 1
```

### PR DESCRIPTION
**Pull request by iabdalkader**
Requires and depends on the following for (WiFi/BT, MSC support and misc fixes.):
- [x] #7561 
- [x] #7698
- [x] #7667  
- [x] #7668 
- [x] #7402
- [x] #7683 
- [x] #7692
- [x] #7694
- [x] #7628
- [x] #7552
- [x] #7503
- [x] micropython/micropython-media#5

Closes #7634

NOTE: Until this patch raspberrypi/pico-sdk#537 gets merged and makes it to a stable release, Arduino Nano RP2040 boards with the newer flash part will Not work.

### CODE REVIEW

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py
Comment (dpgeorge) [inline ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py:3]: Looks like that Arduino driver is LGPL.  It's not possible to put such code in this repo.

Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py:3]: Actually it's mostly based on this: https://github.com/hoihu/projects/blob/master/raspi-hat/lsm9ds1.py which is MIT..I only ported the init sequence from the Arduino driver, let me know if it still can't be added and maybe I can ask them to release another version with a compatible license.

Comment (hoihu) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py:3]: > mostly based on this: https://github.com/hoihu/projects/blob/master/raspi-hat/lsm9ds1.py

Haven't used it for a longer time, should hopefully still work.. no problem to use it anyway, I did the implementation straight from the datasheet at that time.

Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py:3]: @hoihu Yes it's still very much working :) I use it as well with the MIT license in your repo, and lsm6dsox is somewhat based on it (the lsm9ds1), but uses the init sequence from the Arduino library (with reference to both).

**Comment by iabdalkader** [comment]
@dpgeorge I added audio recording/streaming with pdm mic support for this board in the4aee33b6a305a6f965a73177aed9bb7222756317 it's completely self-contained within the board dir. Should it be moved up to the port level ?

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py]: rewrote it.

**Comment by robert-hh** [comment]
I found this set of PR's a little bit confusing and hard for maintainers to deal with it. I wanted to test WiFi support only for the Nano and abandoned it given the set of PRs I found. What I had expected was:
a) a single PR based on a current master, which contains all files and setting for WiFi
b) a single PR based on the PR a) adding basic BT support
c) one or more PRs based on c) adding additional BT services, and finally
d) a PR, moving modnetwork and modusocket to extmod.

**Comment by iabdalkader** [comment]
>  I wanted to test WiFi support only for the Nano
Hi, you can just test our [OpenMV firmware](https://github.com/openmv/openmv/releases/download/v4.1.1/firmware_v4.1.1.zip) it's based on MicroPython 1.15 but it has all of these features enabled for the Nano (and some [examples](https://github.com/openmv/openmv/tree/master/scripts/examples/Arduino/Nano-RP2040))

> I found this set of PR's a little bit confusing and hard for maintainers to deal with it. I wanted to test WiFi support only for the Nano and abandoned it given the set of PRs I found. What I had expected was:
> a) a single PR based on a current master, which contains all files and setting for WiFi

I think multiple smaller independent PRs are much easier to process, but I also don't like to cascade PRs like that because most tend to change a lot and some might get dropped...In this case I thought about moving modnetwork/modusocket to extmod only after this PR, so I sent another PR to move these modules, which I don't know yet if it's going to be merged or not, so I'm also leaving the duplicate files in the networking PR as is until it gets decided.

**Comment by iabdalkader** [comment]
@robert-hh I updated the list of linked PRs, you can see that most of them were independent of each other, if they were cascaded on top of each other they would most likely still be unmerged, and none of these smaller fixes would have made it to 1.17 release (for example the flash fix and MSC support)

**Comment by iabdalkader** [comment]
@dpgeorge Hi, can we get this back on track ? I'm 3 PRs away (including this one) from fully supporting this board in upstream, and the 3 remaining PRs have been ready for too long without any review or feedback.

**Comment by dpgeorge** [comment]
> Hi, can we get this back on track ?

Yes!

I reviewed the main WiFi driver #7668.

As for this PR, I'd really prefer if it did not include the audio driver because it makes it much harder to review, and the discussion will be longer (eg it includes an ST driver that needs to go elsewhere).  Please make this PR just about adding a board definition for the RP2040, then audio can be discussed separately.

**Comment by iabdalkader** [comment]
> I'd really prefer if it did not include the audio driver

Audio driver + modules removed.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py
Comment (dpgeorge) [inline ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py:1]: This file looks like a copy of that in `examples/bluetooth`.  It would be better to include it in the manifest by referring to that file there, instead of copying it.

But note that no other BLE-enabled board includes this file.  What's the reason to include it here?  Related to that, did you see aioble https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble?  That's a very nice, user friendly asyncio-based wrapper around the `bluetooth` module, and maybe it's worth including that instead.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py
Comment (dpgeorge) [inline ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py:1]: This looks like a generic driver and so would be better placed somewhere in `drivers/...`, and then linked via the manifest file.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h
Comment (dpgeorge) [inline ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h:14]: Are these the Arduino VID/PID?  I think they should be the MicroPython ones used in stm32, because with this firmware installed it's not running Arduino anymore... not sure exactly what the correct thing is here.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h
Comment (dpgeorge) [inline ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h:29]: That's a pretty slow baudrate for BLE (also quite an unusual UART baudrate)... can it be increased?

**Comment by dpgeorge** [comment]
Can you please add a `board.json` file for this new board?  See `ports/rp2/boards/PIMORONI_TINY2040/board.json` for an example to copy and adjust.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py:1]: Yes it's an exact copy, and it's frozen so that examples like `examples/bluetooth/ble_temperature.py` work out of the box (without having to somehow copy `ble_advertising.py` to the filesystem). I'll try to add it to the manifest.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h:14]: Yes these are indeed Arduino's VID/PID assigned to this board when running MicroPython, actually when it's running our firmware, but it's based on MicroPython so it will do for now and we may change the PID later. I saw a discussion about that before, and I think the VID/PID should uniquely identify the hardware and software not just the software which can be running on different boards. Knowing that some board is running MicroPython is not very useful on its own, if we can identify the board we can for example update a bootloader or revert its firmware etc.. just my view on the issue.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h:29]: Higher baudrates don't work, I think it must be this specific baudrate or close to it (115200 works).

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py
Comment (jimmo) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py:1]: If you have the flash space I would definitely prefer to include `aioble` instead. For a user of this board, it's a far more useful library, and also specifically for advertising it includes more payload fields.

You can add it to the manifest with

```py
include("$(MPY_LIB_DIR)/micropython/bluetooth/aioble/manifest.py", client=True, central=True, l2cap=True, security=True)
```

Then a user can directly run https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/examples/temp_sensor.py as an alternative to `examples/bluetooth/ble_temperature.py`.

**Comment by iabdalkader** [comment]
> Can you please add a `board.json` file for this new board? See `ports/rp2/boards/PIMORONI_TINY2040/board.json` for an example to copy and adjust.

done.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py]: Can I also add example(s) for the IMU usage ?

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: @jimmo Yes it has 16MBytes of flash, I just have minor concerns about porting another blinky example based on the temp example, but anyway that path doesn't exist ?

Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: I just realized it's in another repo, but how can I link it in the board manifest  ?

Comment (jimmo) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: > I just realized it's in another repo, but how can I link it in the board manifest ?

Exactly as above (I copied it from a board definition I use for testing). The expectation is that the micropython-lib repo is cloned as a sibling directory of the micropython repo (or MPY_LIB_DIR is set otherwise).

The build server is set up like this and e.g. see ports/esp32/boards/manifest_release.py which does something similar for urequests.

Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: Thanks, I managed to test it but it's definitely not as stable as the non-aio one, I get disconnected randomly after a while and can't reconnect. Is this on purpose ? The example seems to be running in a while True.

image

Side note, not important but I noticed it's missing something the indicate property ? which I think allows me to auto-pull the value like with the other ble library/example (using nrf connect for testing, see the screenshot, this is the missing option).

image

Comment (jimmo) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: Interesting... we've been using `aioble` elsewhere (on pybd and stm32-with-external-hci) without issues. I have ordered a Nano 2040 and will investigate when it arrives this week. Does it stop advertising entirely, or just no longer allow connection?

> Side note, not important but I noticed it's missing something the indicate property ?

Yeah, the modbluetooth example sent notifications every 10 seconds. (It has the code to send indications but it's not enabled).

In the `aioble` example, add `True` as a second argument to characteristic.write (i.e. `temp_characteristic.write(_encode_temperature(t), True)`) to automatically notify subscribed clients on write. That should probably be the default in that example, I will send a PR tomorrow.

Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: > Does it stop advertising entirely, or just no longer allow connection?

It just no longer allows connections until I hit ctrl+c and run the script again (paste, ctrl-d etc..)

Comment (dpgeorge) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: We'll definitely fix any issues `aioble` has.  And the goal would be to promote `aioble` and its examples as the recommended way to use bluetooth.  So I think it makes sense to freeze in `aioble` now instead of `ble_advertising.py`.

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py
Comment (dpgeorge) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py]: Yes, please put examples next to the driver, so in `drivers/lsm6dsox/`.

Comment (dpgeorge) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/lsm6dsox.py]: (Eventually would be good to move this and other drivers to `micropython-lib`, and then examples/tests/README would live alongside the driver.)

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h
Comment (dpgeorge) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/mpconfigboard.h:14]: Ok, let's leave them as-is (the PICO actually has custom VID/PID for PICO running MicroPython).

File: ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py
Comment (iabdalkader) [reply ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py]: Okay, switched to aioble.

File: drivers/lsm6dsox/lsm6dsox.py
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:71]: Are these constants intended to be public (part of the API of this class) or private?  If they are private then it's best to prefix them with `_` because then they are fully optimised away and do not contribute at all to the bytecode.  That would save quite a bit of RAM (especially when compiled to .mpy).

File: drivers/lsm6dsox/lsm6dsox.py
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:123]: When optimisations are enabled in mpy-cross (eg -O3) assertions like this are removed from the code.  In such a case there would not be any error checking.

If the intention is to keep these error checks even when opt is enabled, then they need to be changed to: `if not <cond>: raise ValueError(...)`.

File: drivers/lsm6dsox/lsm6dsox.py
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:43]: There should be a `from micropython import const` because const is used below (it's not strictly needed but is good coding style to make it valid Python).

**Review by dpgeorge — CHANGES_REQUESTED** [review]

File: drivers/lsm6dsox/lsm6dsox.py
Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py:71]: Private, prefixed with `_`

**Comment by iabdalkader** [comment]
> It would be great if you could also please split the single commit here into two commits

Done.

File: drivers/lsm6dsox/lsm6dsox.py
Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py:124]: Good point, I missed this will fix.

**Comment by iabdalkader** [comment]
There's 1 PR left (the MSC support) it would be nice if it's merged first, note it's not enabled for the port just for the Nano. It's currently needed for the IMU MLC example and later for Audio, which I will send in a following PR after everything is merged and after some refactoring.

**Comment by dpgeorge** [comment]
> There's 1 PR left (the MSC support) it would be nice if it's merged first, note it's not enabled for the port just for the Nano.

But if MSC is enabled for the Nano then this board is now different to all other rp2 boards.  And that may lead to confusion, and questions about why it's enabled on Nano only, and the only way to solve that would be to enable MSC on all rp2 boards.  It's a bit of a slippery slope.

> It's currently needed for the IMU MLC example and later for Audio

It's possible to use `mpremote` to copy files to/from the board.  Yes there is some minor set up required to get `mpremote` working (make sure Python is installed, then `pip install mpremote`), but having `mpremote` installed gives a lot of other useful functionality.  Once you have your workflow based on `mpremote` you forget all about MSC.

-----

From the discussion in #7402 there are good points for and against MSC, but it's still not clear whether MSC should be supported on rp2.  And as I say above I think it'll be confusing if some rp2 boards support it and some don't.

I would rather merge this PR now, so it's not blocked by the MSC discussion, and continue the MSC discussion in #7402.

**Comment by iabdalkader** [comment]
> I would rather merge this PR now, so it's not blocked by the MSC discussion

I agree on that let's revisit later.

**Comment by dpgeorge** [comment]
Ok, merged in f082793ac44dac77517bab3361a80f062b500345 and c3dceb1c3223012ae29144182d14c19678005e05 !

----

I tested it and it works well.  But the results returned from WLAN scan do not have the same format as other WLAN drivers, that will need to be fixed.

**Comment by iabdalkader** [comment]
Great thanks! 

> that will need to be fixed

I will check this and send a PR.

**Comment by iabdalkader** [comment]
This should be closed  #7634

**Comment by iabdalkader** [comment]
>  But the results returned from WLAN scan do not have the same format as

Question, does it have to match the data types too ? because I return a nicely formatted mac address (vs just bytes).

**Comment by dpgeorge** [comment]
> does it have to match the data types too

Yes.  Otherwise code will not be compatible across devices.  Eg I have a script that does wifi scanning and prints out a nice result where the MAC is formatted, and if the MAC is already formatted then it will get double formatted.

**Comment by iabdalkader** [comment]
> Yes. Otherwise code will not be compatible across devices.

It's now the same order/data type as other NICs, plus small fix to netinfo command #8004

**Pull request closed without merging**
